service/controller/clusterapi/cluster.go

```diff
@@ -0,0 +1,115 @@
+	clusterv1alpha1 "github.com/giantswarm/apiextensions/pkg/apis/cluster/v1alpha1"
+	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/client/k8scrdclient"
+	"github.com/giantswarm/operatorkit/controller"
+	"github.com/giantswarm/operatorkit/informer"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+
+	"github.com/giantswarm/cluster-operator/pkg/cluster"
+	v16 "github.com/giantswarm/cluster-operator/service/controller/clusterapi/v16"
+)
+
+// ClusterConfig contains necessary dependencies and settings for
+// Cluster API's Cluster CRD controller implementation.
+type ClusterConfig struct {
+	BaseClusterConfig *cluster.Config
+	CMAClient         clientset.Interface
+	G8sClient         versioned.Interface
+	K8sExtClient      apiextensionsclient.Interface
+	Logger            micrologger.Logger
+
+	ProjectName string
+}
+
+type Cluster struct {
+	*controller.Controller
+}
+
+// NewCluster returns a configured AWSClusterConfig controller implementation.
+func NewCluster(config ClusterConfig) (*Cluster, error) {
+	if config.G8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.G8sClient must not be empty", config)
+	}
+
+	var err error
+
+	var crdClient *k8scrdclient.CRDClient
+	{
+		c := k8scrdclient.Config{
+			K8sExtClient: config.K8sExtClient,
+			Logger:       config.Logger,
+		}
+
+		crdClient, err = k8scrdclient.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var newInformer *informer.Informer
+	{
+		c := informer.Config{
+			Logger:  config.Logger,
+			Watcher: config.CMAClient.ClusterV1alpha1().Clusters(corev1.NamespaceAll),
+
+			RateWait:     informer.DefaultRateWait,
+			ResyncPeriod: informer.DefaultResyncPeriod,
+		}
+
+		newInformer, err = informer.New(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var resourceSetV16 *controller.ResourceSet
+	{
+		c := v16.ResourceSetConfig{
+			BaseClusterConfig: config.BaseClusterConfig,
+			CMAClient:         config.CMAClient,
+			G8sClient:         config.G8sClient,
+			Logger:            config.Logger,
+		}
+
+		resourceSetV16, err = v16.NewResourceSet(c)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var clusterController *controller.Controller
+	{
+		c := controller.Config{
+			CRD:       clusterv1alpha1.NewClusterCRD(),
+			CRDClient: crdClient,
+			Informer:  newInformer,
+			Logger:    config.Logger,
+			ResourceSets: []*controller.ResourceSet{
+				resourceSetV16,
+			},
+			RESTClient: config.CMAClient.ClusterV1alpha1().RESTClient(),
+
+			// Name is used to compute finalizer names. This here results in something
+			// like operatorkit.giantswarm.io/aws-operator-cluster-controller.
+			Name: config.ProjectName + "-cluster-controller",
```

Comment:
I was pondering on this. Is this actually needed in `cluster-operator` case? Could it be just `config.ProjectName` ?

Reply:
After thinking about this. This is for finalizers etc. in Cluster CR. Using just `ProjectName` makes sense here.